app/state_machines/allocation_state_machine.rb

```diff
@@ -0,0 +1,11 @@
+class AllocationStateMachine < FiniteMachine::Definition
+  initial :unfilled
+
+  event :fill, %i[none unfilled filled] => :filled
+  event :unfill, %i[none unfilled filled] => :unfilled
+  event :cancel, %i[none unfilled filled] => :cancelled
```

Comment:
It doesn't feel idiomatic to have the transitions back to the same state - that would make the graph look kinda weird. This feels cleaner:

```ruby
event :fill, %i[none unfilled] => :filled
event :unfill, %i[none filled] => :unfilled
event :cancel, %i[none unfilled filled] => :cancelled
```

Reply:
Also might be worth adding a terminal state for `cancelled` as there's no way to transition out of that. Simply:

```ruby
terminal :cancelled
```